// react-selectize/src-babel/ReactSelectize.js
import React from "react"
import ReactDOM from "react-dom"

module.exports = class ReactSelectize extends React.Component {

    constructor(props) {
        super(props)
    }

    render() {
        return (
            <div className="ReactSelectize">
                {this.renderHiddenInputElement()}
            </div>
        )
    }

    renderDropdownMenu() {}

    renderHiddenInputElement() {
        if (this.props.name) {
            return <input 
                type = "hidden" 
                name = {this.props.name} 
                value = {this.props.serialize(this.props.values)}
            />
        } else {
            return null
        }
    }

    renderResetButton() {
        return (
            <div className = "ReactSelectize-resetButtonContainer" onClick = {(e) => {
                this.props.onValuesChange([], () => this.props.onSearchChange("", () => this.highlightAndFocus()))
                return cancelEvent(e)
            }}>
                {this.props.renderResetButton()}
            </div>
        )
    }

    renderSearchField() {}

    renderSelectedValues() {}

    renderToggleButton() {
        return (
            <div className = "ReactSelectize-toggleButtonContainer" onMouseDown = {(e) => {
                if (this.props.open) {
                    this.closeDropdownMenu()
                } else {
                    this.props.onAnchorChange(_.last(this.props.values), () => this.openDropdownMenu())
                }
                return cancelEvent(e)
            }}>
                {this.props.renderToggleButton({
                    open: this.props.open, 
                    dropdownDirection: this.props.dropdownDirection
                })}
            </div>
        )
    }

    closeDropdownMenu(callback) {
        this.props.onOpenChange(false, callback)
    }

    openDropdownMenu(callback) {
        this.props.onOpenChange(!this.props.disabled, callback)
    }

}
